File: AdSec/.NET/ReinforcedConcreteAnalysis/ReinforcedConcreteAnalysis.cs
using System;
using Oasys.AdSec;
using Oasys.AdSec.DesignCode;
using Oasys.AdSec.Materials;
using Oasys.AdSec.StandardMaterials;
using Oasys.Profiles;
using Oasys.AdSec.Reinforcement;
using Oasys.AdSec.Reinforcement.Groups;
using Oasys.AdSec.Reinforcement.Layers;
using OasysUnits;

namespace ReinforcedConcreteAnalysis
{
    /// <summary>
    /// This example shows how to define a reinforced section and analyse it.
    /// The example goes on to apply a load and check the utilisation and
    /// maximum crack width.
    ///
    /// You might like to run the 'ApiVersion' example first, just to check
    /// that the API is installed correctly.
    /// </summary>
    public static class ReinforcedConcreteAnalysis
    {
        public static void Main()
        {
            // Create a rectangular section
            var profile = IRectangleProfile.Create(
                Length.FromMillimeters(800),
                Length.FromMillimeters(400)
            );
            IConcrete sectionMaterial = Concrete
                .EN1992
                .Part1_1
                .Edition_2004
                .NationalAnnex
                .GB
                .Edition_2014
                .C40_50;
            var section = ISection.Create(profile, sectionMaterial);

            // Set the cover
            section.Cover = ICover.Create(Length.FromMillimeters(40));

            // Set some reinforcement
            IReinforcement reinforcementMaterial = Reinforcement
                .Steel
                .EN1992
                .Part1_1
                .Edition_2004
                .NationalAnnex
                .GB
                .Edition_2014
                .S500B;
            IBarBundle bar20mm = IBarBundle.Create(
                reinforcementMaterial,
                Length.FromMillimeters(20)
            );
            IBarBundle bar16mm = IBarBundle.Create(
                reinforcementMaterial,
                Length.FromMillimeters(16)
            );
            IBarBundle bar12mm = IBarBundle.Create(
                reinforcementMaterial,
                Length.FromMillimeters(12)
            );

            // Define top reinforcement
            ILayer topLayer = ILayerByBarCount.Create(4, bar16mm);
            ITemplateGroup topReinforcement = ITemplateGroup.Create(ITemplateGroup.Face.Top);
            topReinforcement.Layers.Add(topLayer);

            // Define bottom reinforcement
            ILayer bottomLayerOne = ILayerByBarCount.Create(4, bar20mm);
            ILayer bottomLayerTwo = ILayerByBarCount.Create(4, bar16mm);
            ITemplateGroup bottomReinforcement = ITemplateGroup.Create(ITemplateGroup.Face.Bottom);
            bottomReinforcement.Layers.Add(bottomLayerOne);
            bottomReinforcement.Layers.Add(bottomLayerTwo);

            // Define link (stirrup)
            ILinkGroup link = ILinkGroup.Create(bar12mm);

            // Add defined reinforcement groups to section
            section.ReinforcementGroups.Add(topReinforcement);
            section.ReinforcementGroups.Add(bottomReinforcement);
            section.ReinforcementGroups.Add(link);

            // Analyse the section to create a solution
            var adSec = IAdSec.Create(EN1992.Part1_1.Edition_2004.NationalAnnex.GB.Edition_2014);
            ISolution solution = adSec.Analyse(section);

            // Calculate utilisation for a particular load
            var axialForce = Force.FromKilonewtons(-100);
            var majorAxisBending = Moment.FromKilonewtonMeters(-500);
            var minorAxisBending = Moment.Zero;
            var load = ILoad.Create(axialForce, majorAxisBending, minorAxisBending);
            IStrengthResult strengthResult = solution.Strength.Check(load);

            // Display utilisation as a percentage
            double utilisation = Math.Round(strengthResult.LoadUtilisation.Percent, 1);
            Console.WriteLine($"The utilisation is: {utilisation}%");

            // Calculate the serviceability crack width under the same load
            IServiceabilityResult serviceabilityResult = solution.Serviceability.Check(load);

            // Display the crack width in mm
            double crackWidth = Math.Round(
                serviceabilityResult.MaximumWidthCrack.Width.Millimeters,
                2
            );
            Console.WriteLine($"The maximum crack width is: {crackWidth}mm");
        }
    }
}

File: AdSec/Python/ReinforcedConcreteAnalysis.py
# Load the AdSec API

# Import modules from namespace
from Oasys.AdSec import IAdSec, ILoad, ISection
from Oasys.AdSec.DesignCode import EN1992
from Oasys.AdSec.Reinforcement import IBarBundle, ICover
from Oasys.AdSec.Reinforcement.Groups import ILinkGroup, ITemplateGroup
from Oasys.AdSec.Reinforcement.Layers import ILayerByBarCount
from Oasys.AdSec.StandardMaterials import Concrete, Reinforcement
from Oasys.Profiles import IRectangleProfile
from OasysUnits import Force, Length, Moment
from OasysUnits.Units import ForceUnit, LengthUnit, MomentUnit


# This example shows how to define a reinforced section and analyse it.
# The example goes on to apply a load and check the utilisation and
# maximum crack width.
#
# You might like to run the 'version.py' example first, just to check
# that the API is installed correctly.
def strength_analysis():
    # Create a rectangular section
    depth = Length(float(800), LengthUnit.Millimeter)
    width = Length(float(400), LengthUnit.Millimeter)
    profile = IRectangleProfile.Create(depth, width)
    section_material = Concrete.EN1992.Part1_1.Edition_2004.NationalAnnex.GB.Edition_2014.C40_50
    section = ISection.Create(profile, section_material)

    # Set the cover
    section.Cover = ICover.Create(Length(float(40), LengthUnit.Millimeter))

    # Set some reinforcement
    reinforcement_material = Reinforcement.Steel.EN1992.Part1_1.Edition_2004.NationalAnnex.GB.Edition_2014.S500B
    bar20mm = IBarBundle.Create(reinforcement_material, Length(float(20), LengthUnit.Millimeter))
    bar16mm = IBarBundle.Create(reinforcement_material, Length(float(16), LengthUnit.Millimeter))
    bar12mm = IBarBundle.Create(reinforcement_material, Length(float(12), LengthUnit.Millimeter))

    # Define top reinforcement
    layer_for_top_group = ILayerByBarCount.Create(4, bar16mm)
    top_group = ITemplateGroup.Create(ITemplateGroup.Face.Top)
    top_group.Layers.Add(layer_for_top_group)

    # Define bottom reinforcement
    layer_one_for_bottom_group = ILayerByBarCount.Create(4, bar20mm)
    layer_two_for_bottom_group = ILayerByBarCount.Create(4, bar16mm)
    bottom_group = ITemplateGroup.Create(ITemplateGroup.Face.Bottom)
    bottom_group.Layers.Add(layer_one_for_bottom_group)
    bottom_group.Layers.Add(layer_two_for_bottom_group)

    # Define link (stirrup)
    link = ILinkGroup.Create(bar12mm)

    # Add defined reinforcement groups to section
    section.ReinforcementGroups.Add(top_group)
    section.ReinforcementGroups.Add(bottom_group)
    section.ReinforcementGroups.Add(link)

    # Analyse the section to create a solution
    ad_sec = IAdSec.Create(EN1992.Part1_1.Edition_2004.NationalAnnex.GB.Edition_2014)
    solution = ad_sec.Analyse(section)

    # Calculate utilisation for a particular load
    axial_force = Force(float(-100), ForceUnit.Kilonewton)
    major_axis_bending = Moment(float(-500), MomentUnit.KilonewtonMeter)
    minor_axis_bending = Moment.Zero
    load = ILoad.Create(axial_force, major_axis_bending, minor_axis_bending)
    strength_result = solution.Strength.Check(load)

    # Display utilisation as a percentage
    utilisation = round(strength_result.LoadUtilisation.Percent, 1)
    print("The utilisation is: " + str(utilisation) + "%")

    # Calculate the serviceability crack width under the same load
    serviceability_result = solution.Serviceability.Check(load)

    # Display the crack width in mm
    crack_width = round(serviceability_result.MaximumWidthCrack.Width.Millimeters, 2)
    print("The maximum crack width is: " + str(crack_width) + " mm")


if __name__ == "__main__":
    strength_analysis()
